lxd/response/smart.go

```diff
@@ -9,51 +9,76 @@ import (
 	"github.com/canonical/lxd/shared/api"
 )
 
+// ErrFunc represents the signature of additional functions that can be passed to Init. This allows deeper error
+// inspections using e.g. [errors.As]. SmartError will evaluate each function in errFuncs, if any return a valid HTTP
+// status code, then that status code will be used (otherwise [http.StatusInternalServerError] is returned).
+type ErrFunc func(error) int
+
+// httpResponseErrors contains some default mappings of HTTP status codes to sentinel errors that are defined in the Go
+// standard library. Additional errors can be added to this map on a call to Init.
 var httpResponseErrors = map[int][]error{
 	http.StatusNotFound:  {os.ErrNotExist, sql.ErrNoRows},
 	http.StatusForbidden: {os.ErrPermission},
 }
 
-// SmartError returns the right error message based on err.
-// It uses the stdlib errors package to unwrap the error and find the cause.
+// defaultErrFunc checks all sentinel errors in httpResponseErrors and returns the appropriate code if a match is found.
+// Otherwise, it returns zero (which is not a valid HTTP status code, and so indicates that there was no match).
+func defaultErrFunc(err error) int {
+	for code, errList := range httpResponseErrors {
+		for _, sentinelErr := range errList {
+			if errors.Is(err, sentinelErr) {
+				return code
+			}
+		}
+	}
+
+	return 0
+}
+
+// errFuncs is a slice of ErrFunc. SmartError iterates over this list to determine the appropriate HTTP response code
+// for the given error.
+var errFuncs = []ErrFunc{
+	defaultErrFunc,
+}
+
+// SmartError returns a Response with a code extracted from the given error.
 func SmartError(err error) Response {
-	if err == nil {
+	code := getCode(err)
+	if code == http.StatusOK {
 		return EmptySyncResponse
 	}
 
+	return &errorResponse{code: code, err: err}
+}
+
+// getCode gets HTTP response code for a given error, or returns [http.StatusInternalServerError].
+func getCode(err error) int {
+	// If no error, return 200 OK.
+	if err == nil {
+		return http.StatusOK
+	}
+
+	// Check if it is an [api.StatusError] and if so, return the associated code.
 	statusCode, found := api.StatusErrorMatch(err)
 	if found {
-		return &errorResponse{statusCode, err}
+		return statusCode
 	}
 
-	for httpStatusCode, checkErrs := range httpResponseErrors {
-		for _, checkErr := range checkErrs {
-			if errors.Is(err, checkErr) {
-				if err != checkErr {
-					// If the error has been wrapped return the top-level error message.
-					return &errorResponse{httpStatusCode, err}
-				}
-
-				// If the error hasn't been wrapped, use a generic error.
-				return &errorResponse{httpStatusCode, nil}
```

Comment:
@roosterfish it turns out we do still need the map response errors. I had missed this logic in my refactor, which returns the generic HTTP status text when a sentinel error is encountered and hasn't been wrapped.

Reply:
Thanks, I've made one more change to make it more flexible:

The initialisation and handling of sentinel errors is kept as is. Now we'll have a list of functions with `func(err) (code int, msg string)` allowing the registered error matchers to overwrite the error message sent over the API. This could be useful if we figure out a way to automatically map constraint violations to API friendly messages.